Added Buy on PacerDash button for RECAP docs

albertisfu opened the pull request:
Added the new `Buy on PacerDash` button for RECAP documents.

Some comments to confirm if they are correct: 

- The Install RECAP extension modal is only shown on the "Buy on PACER" button, on the "Buy on PacerDash" button we open the link directly.

- Is it possible that a document doesn't have a `pacer_doc_id` since we need it to construct the PacerDash URL, in case we don't have it, should we hide the PacerDash button? or there is another way to construct the link for documents without a `pacer_doc_id`? 

 Here are some screenshots of what it looks like for different cases:

**Document not available, price available:**
![document_no_available_price_](https://user-images.githubusercontent.com/486004/170392097-944086cc-b9d7-4fb6-b22a-69daba7307c3.png)

**Document not available, price not available:**
![document_no_available_no_price_*](https://user-images.githubusercontent.com/486004/170392320-d05d50e0-c26b-4478-affb-98fae8b47d23.png)

**Document not available, free on PACER:**
![document_no_available_free_on_pacer_](https://user-images.githubusercontent.com/486004/170392350-13493c91-212c-4dbf-a3ef-9a45db72a7f3.png)

**Document not available, mobile:**
![document_no_available_mobile_](https://user-images.githubusercontent.com/486004/170392426-cd1964a7-055e-4bfa-87c3-bafae2a9fd80.png)

**Document available, no free on PACER, price available**,  the horizontal line between the download options and the "Buy" options is preserved as required:
![document_available_no_free_on_pacer_price_](https://user-images.githubusercontent.com/486004/170391452-51a9d869-e45f-44fa-b99d-e5757bdf3efb.png)

**Document available, no free on PACER, no price available**:
![document_available_no_free_on_pacer_no_price_](https://user-images.githubusercontent.com/486004/170391648-255985de-530e-4f76-b2a4-f6390d9dea2c.png)

**Document available, free on PACER**, here we don't show the "Buy" buttons as required:
![document_available_free_on_pacer_](https://user-images.githubusercontent.com/486004/170391748-c72b8d03-86a5-4b67-a752-c438ba2a5d68.png)

**Document available, mobile:**
![documen_available_mobile_](https://user-images.githubusercontent.com/486004/170392055-6d7bcedb-e96b-4b9b-8e14-b93535c44014.png)

Also added on RECAP document detail:

**Document available**
![recap_document_available_](https://user-images.githubusercontent.com/486004/170392469-5e6d0509-a165-4494-899f-00b523995536.png)

**Document not available**
![recap_document_no_available_](https://user-images.githubusercontent.com/486004/170392515-672d96ff-9a30-4fb8-a71d-af3675bf58ae.png)

albertisfu commented:
there is that Semgrep suggestion of not using `{{ rd.pacerdash_url }} ` and instead using a `{% url  %}` tag, however, I'm not sure if this could apply to us since we are building an external URL.

quinnr commented:
> there is that Semgrep suggestion of not using `{{ rd.pacerdash_url }} ` and instead using a `{% url %}` tag, however, I'm not sure if this could apply to us since we are building an external URL.

Looking through the documentation on the alert and `{% url %}` tag documentation I think you're right as the url tag won't point to external websites. Since the inputs aren't built on any input generated by users I think the security risk is relatively small, although it couldn't hurt to sanitize the input to ensure `pacer_doc_id` is in the expected format if that does not already happen at the database level. 

Looks like Django has some [sitewide settings](https://www.stackhawk.com/blog/django-xss-examples-prevention/) to address this security concern as well in addition to the Content Security Policy header mentioned in the semgrep alert.

albertisfu commented:
Thank you! @quinnr I've checked and yes it seems to be safe since Django automatically scapes some HTML characters. 

For instance a pacer_doc_id like this:
`"<script>alert('XSS');</script>"`
Is rendered in the URL as follows:
`"https://www.pacerdash.com/court-listener-checkouts/&quot;<script>alert('XSS');</script>&quot;"`

mlissner commented:
Just looking at the screenshots and your questions, Alberto, and then @flooie, if you can do review, that'd be great:

1. If we don't have the pacer_doc_id, then we can sometimes construct the link to PACER, but we can't do one to PacerDash, so no link in that case. 

2. You are correct that we don't show the Install RECAP modal for people that go to PacerDash.

3. The right edge of the drop down button itself should align on the right with the drop down arrow. I'm guessing the alignment is off because you shrank it? If it's hard to get right, we can just not shrink it and keep our lives simpler.

4. The buttons in your last screenshot look funky, and anyway should be combined into a drop down.

5. Semgrep is fine. We trust the data source.

6. I was wrong when I thought we could avoid putting a link to PACER when the docs are free — people will probably still want that from time to time. Let's continue not showing a link to PacerDash for those docs, but let's show the link to PACER. If it's easy to do, it might be good for it to say "From PACER" instead of "Buy on PACER" when it's free, so it'd say:

   - Download PDF v
       - From CourtListener
       - From Internet Archive
       - Maybe a horizontal line; your call when you see it.
       - From PACER ($0.00)

    If that's a pain to tweak, we can just go back to how it was.

mlissner reviewed (changes requested):
I skimmed the code, but I'll let @flooie do a full review. It's what I'd expect, at a glance.

albertisfu commented:
Great, I'll work on these changes, thanks!

albertisfu commented:
These changes are ready:

- If we don't have a pacer_doc_id we don't show the PacerDash button in any case.
![document_no_available_no_pacer_id_price](https://user-images.githubusercontent.com/486004/170615020-2c10591f-e444-4f3b-95ee-ca38e90bda58.png)

- The dropdown PacerDash button was adjusted and aligned, now is the same size as the main button.
![document_no_available_price_](https://user-images.githubusercontent.com/486004/170615071-f34a49f4-9b31-4d17-bcd7-3d2c340c612a.png)
Also works with the smaller button (no price available)
![document_no_available_no_price_](https://user-images.githubusercontent.com/486004/170615217-43fed55b-8d24-4c75-be90-5585c7a4e45e.png)

- For the RECAP document detail page, added the PacerDash button as a dropdown:
![recap_document_no_available_](https://user-images.githubusercontent.com/486004/170615349-2e705bcf-6345-446f-aca1-29fe18c23cb9.png)

- We keep the PACER button when we have the document and it's free, changed the text to "From PACER ($0.00)", I think the horizontal line is good since it helps to differentiate the direct download links from the one from PACER, here we don't show the PacerDash button.
![document_available_free_on_pacer_](https://user-images.githubusercontent.com/486004/170615496-ae7ff14f-13d9-4b8b-94d8-afbcfd21e7b0.png)

mlissner commented:
In the second to last image, the icon in the button needs a space before it, I think. Otherwise, looks great, thanks. @flooie, to you for code review, please. :)

mlissner commented:
Looks like this got stuck. @flooie can you please do a review here?

flooie commented on cl/search/models.py:
```diff
@@ -1186,6 +1186,17 @@ def pacer_url(self):
                 )
 
+    @property
+    def pacerdash_url(self) -> str:
+        """Construct the PacerDash URL, if we can. Else, return None."""
+
+        if self.pacer_doc_id:
+            return (
+                "https://www.pacerdash.com/court-listener-checkouts/"
```
Two small quibbles. 

One I think this could use better docstrings.

two.  I think you could return None and simplify the method.  

I may be wrong but I would do this 

```
@property
    def pacerdash_url(self) -> Optional[str]:
        """Construct the PacerDash URL, if we can. Else, return None."""

        if not self.pacer_doc_id: return None
        return (
            "https://www.pacerdash.com/court-listener-checkouts/"
            f"{self.pacer_doc_id}"
        )
```

albertisfu replied:
Improved docstring and try to change the return None into one line `if not self.pacer_doc_id: return None` as suggested but seems that `black` didn't like it because it reformatted into two lines.
```
if not self.pacer_doc_id:
     return None
```

but it works!

flooie commented on cl/custom_filters/templatetags/pacer.py:
```diff
@@ -14,3 +14,15 @@ def price(rd: RECAPDocument) -> str:
+
+
+@register.filter
+def pacerdash_price(rd: RECAPDocument) -> str:
+    if rd.is_free_on_pacer:
+        return "0.00"
+
```
Docstrings missing...

And I think they are incorrect.

flooie commented on cl/custom_filters/templatetags/pacer.py:
```diff
@@ -14,3 +14,15 @@ def price(rd: RECAPDocument) -> str:
+
+@register.filter
+def pacerdash_price(rd: RECAPDocument) -> str:
+    if rd.is_free_on_pacer:
+        return "0.00"
```
is this necessary.

flooie replied:
ok.

flooie commented on cl/opinion_page/templates/includes/de_list.html:
```diff
@@ -88,46 +88,63 @@
+                                {% endif %}
+                                target="_blank"
+                                rel="nofollow">
+                                  {% if rd|price != "0.00" %}Buy on PACER{% else %}From PACER{% endif %} {% if rd.page_count %}(${{ rd|price }}){% endif %}
+                              </a>
+                          </li>
+                          {% if rd.pacerdash_url and rd|price != "0.00" %}
```
instead of `rd|price != 0.00` ... can we do `rd.is_free_on_pacer`

flooie commented:
Thanks for waiting me out @albertisfu  and @mlissner this looks pretty good to me.  Just a couple of things to tweak I think and this should be good to go.

albertisfu commented:
Thank you @flooie for your revision! I'll work on these changes and suggestions!

albertisfu commented on cl/custom_filters/templatetags/pacer.py:
```diff
@@ -14,3 +14,15 @@ def price(rd: RECAPDocument) -> str:
+        cost = rd.page_count * 0.10
+        pacerdash_cost = (min(3, cost) * 1.029) + 0.90
+        return f"{pacerdash_cost:.2f}"
+    return ""
```
done, also added docstring to `def price` that it's a very similar method

flooie replied:
great.

albertisfu commented:
@flooie I've applied these changes, let me know what you think, I wrote some comments above.
Thank you!

flooie commented on cl/opinion_page/templates/includes/de_list.html:
```diff
@@ -148,24 +165,17 @@
-
-                        {% endif %}
-
+                          title="Buy on PACER {% if rd.page_count %}(${{ rd|price }}){% endif %}"><i class="fa fa-download"></i>
```
This if statement isn't needed right?

flooie replied:
or I see we need it for the ($  )?

albertisfu replied:
yeah, it's used to only show the price inside ($ ) if we have the `page_count`

flooie commented on cl/opinion_page/templates/includes/rd_download_button.html:
```diff
@@ -27,17 +27,18 @@
+             {% endif %}
+              target="_blank"
+              rel="nofollow">Buy on PACER</a>
+            {% if rd.pacerdash_url %}
+              <a href="{{ rd.pacerdash_url }}"
+              target="_blank"
+              rel="nofollow">Buy on PacerDash</a>
+            {% endif %}
```
This comment is more for @mlissner but wouldnt this all be nicer if we built the html link at the filter level with something like 

```
@register.filter
def pacerdash_url(rd: RECAPDocument) -> str:
    """"""
    if not rd.pacerdash_url:
        return ""
    return mark_safe(
        f'<a href="{ rd.pacerdash_url }" target="_blank" rel="nofollow">Buy on PacerDash</a>'
    )
```

and then we would just pass 

`{% pacerdash_url %}`

into the html template and really simplify the HTML

mlissner commented:
This looks great. Thank you both. Now all we need to do is deploy it and write the blog post!
